### ISSUE TITLE
Occasionally, the wrong REPL type is selected.

### ISSUE THREAD

**Issue by austinhaas** [clojure-emacs/inf-clojure#151]
Occasionally, inf-clojure will select the wrong REPL type. The problem appears to start here at inf-clojure--detect-repl-type: https://github.com/clojure-emacs/inf-clojure/blob/master/inf-clojure.el#L297

I added debug messages around inf-clojure--process-reponse-match-p ( https://github.com/clojure-emacs/inf-clojure/blob/master/inf-clojure.el#L1277) and it looks like what is happening is that one of the calls to inf-clojure--process-response, to check if a certain namespace, like planck.repl, is available, will return a blank string, and then the corresponding type will be selected because it isn't nil.

The most obvious symptom of this issue is that the eldoc output will either be absent or show part of an error message. I added a message to print out the type as soon as it was selected, but you can also see it calling the wrong functions if you follow the instructions for logging process activity here: https://github.com/clojure-emacs/inf-clojure#log-process-activity.

To test this, I had to `M-: (setq-local inf-clojure-repl-type nil) RET` in the buffer that I was testing in order for inf-clojure to reattempt to determine the REPL type.

Initially, I thought I could reproduce the problem by evaluating the buffer after the REPL starts, but before triggering eldoc (by moving the cursor inside an expression), but I haven't been able to trigger the issue consistently (only about 3 out of 25 attempts).

**Comment by arichiardi** [comment]
The repl detection part is not tested and I am not surprised this happens. Maybe at some point I had empty string detection but I removed it for ...reasons...

Character based interaction is not good and imho should be at some point deprecated in `inf-clojure` - only socket REPL and in the future pREPL are reliable and linear enough for us to spend time on it.

**Comment by arichiardi** [comment]
Having said that, this is a bug so you caught it. Nice!

**Comment by bbatsov** [comment]
> Character based interaction is not good and imho should be at some point deprecated in inf-clojure - only socket REPL and in the future pREPL are reliable and linear enough for us to spend time on it.

It's at the heart of comint, so I doubt we can ever deprecate it, but we should certainly promote the usage of the socket REPL (or pREPL).

**Comment by sogaiu** [comment]
Regarding inf-clojure-detect-repl-type, I found the following in inf-clojure.el:

```emacs-lisp
(defvar-local inf-clojure-repl-type nil
  "Symbol to define your REPL type.
Its root binding is nil and it can be further customized using
either `setq-local` or an entry in `.dir-locals.el`." )
```
I wanted to try the setq-local approach and am currently trying the following in an initialization file:

```emacs-lisp
(defun arcadia-set-repl-type ()
  (setq-local inf-clojure-repl-type 'clojure))

(add-hook 'inf-clojure-mode-hook #'arcadia-set-repl-type)
```

So far it seems to be working -- i.e. I haven't yet observed any attempts at detection via the sending of forms.

Does that seem appropriate?

**Comment by arichiardi** [comment]
Yes that var will take precedence over the detection machinery so it is probably good to set it in case things go wrong.

**Comment by sogaiu** [comment]
Thanks for the explanation :)

**Comment by charignon** [comment]
Here is how to repro the issue:
```
lein new app testapp
# open testapp/src/core.clj with emacs
# start inf-clojure (which starts lein repl)
# start inf-clojure-minor-mode
# kill the *inf-clojure* buffer 
# disable inf-clojure-minor-mode
cd testapp && lumo -n 5555
# connect to the lumo repl with inf-clojure-connect 
# start inf-clojure-minor-mode
```
At this point the repl type is set to `clojure` instead of `lumo` and things break unless you forcibly force recomputation of the type (e.g. by evaluating `(setq inf-clojure-repl-type nil`)).

I think either of the following two approaches would solve the repro case I showed above:
1/ force recomputation of the repl type whenever we connect to a repl or start a new repl 
2/ set `inf-clojure-repl-type` to nil when someone leaves the minor mode

What do you think @arichiardi ?
